pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/ServerRequestPlanVisitor.java

```diff
@@ -0,0 +1,363 @@
+    visitChildren(node, context);
+    context.getPinotQuery().setSelectList(CalciteRexExpressionParser.overwriteSelectList(
+        node.getProjects(), context.getPinotQuery()));
+    return _aVoid;
+  }
+
+  @Override
+  public Void visitSort(SortNode node, ServerPlanRequestContext context) {
+    visitChildren(node, context);
+    if (node.getCollationKeys().size() > 0) {
+      context.getPinotQuery().setOrderByList(CalciteRexExpressionParser.convertOrderByList(node.getCollationKeys(),
+          node.getCollationDirections(), context.getPinotQuery()));
+    }
+    if (node.getFetch() > 0) {
+      context.getPinotQuery().setLimit(node.getFetch());
+    }
+    if (node.getOffset() > 0) {
+      context.getPinotQuery().setOffset(node.getOffset());
+    }
+    return _aVoid;
+  }
+
+  @Override
+  public Void visitTableScan(TableScanNode node, ServerPlanRequestContext context) {
+    DataSource dataSource = new DataSource();
+    String tableNameWithType = TableNameBuilder.forType(context.getTableType())
+        .tableNameWithType(TableNameBuilder.extractRawTableName(node.getTableName()));
+    dataSource.setTableName(tableNameWithType);
+    context.getPinotQuery().setDataSource(dataSource);
+    context.getPinotQuery().setSelectList(node.getTableScanColumns().stream()
+        .map(RequestUtils::getIdentifierExpression).collect(Collectors.toList()));
+    return _aVoid;
+  }
+
+  @Override
+  public Void visitValue(ValueNode node, ServerPlanRequestContext context) {
+    visitChildren(node, context);
+    return _aVoid;
+  }
+
+  private void visitChildren(StageNode node, ServerPlanRequestContext context) {
+    for (StageNode child : node.getInputs()) {
+      child.visit(this, context);
+    }
+  }
+
+  /**
+   * Helper method to attach the dynamic filter to the given PinotQuery.
+   */
+  private static void attachDynamicFilter(PinotQuery pinotQuery, JoinNode.JoinKeys joinKeys,
+      TransferableBlock dataBlock) {
+    FieldSelectionKeySelector leftSelector = (FieldSelectionKeySelector) joinKeys.getLeftJoinKeySelector();
+    FieldSelectionKeySelector rightSelector = (FieldSelectionKeySelector) joinKeys.getRightJoinKeySelector();
+    List<Expression> expressions = new ArrayList<>();
+    for (int i = 0; i < leftSelector.getColumnIndices().size(); i++) {
+      Expression leftExpr = pinotQuery.getSelectList().get(leftSelector.getColumnIndices().get(i));
+      int rightIdx = rightSelector.getColumnIndices().get(i);
+      Expression dynamicFilter = RequestUtils.getFunctionExpression(TransformFunctionType.INIDSET.name());
+      dynamicFilter.getFunctionCall().setOperands(Arrays.asList(leftExpr, computeIdSet(dataBlock, rightIdx)));
+      Expression filterExpr = RequestUtils.getFunctionExpression(FilterKind.EQUALS.name());
+      filterExpr.getFunctionCall().setOperands(Arrays.asList(dynamicFilter, RequestUtils.getLiteralExpression(true)));
+      expressions.add(filterExpr);
+    }
```

Comment:
here is the conversion to filter expression

- not the best to use IDSET b/c we can't correlated with the row selected out in the right table
- but using IDSET makes it fast.

Reply:
> not the best to use IDSET b/c we can't correlated with the row selected out in the right table

not sure I understand this? what do you mean?

Reply:
for example: in the situation of `SELECT a.col1, b.col2 FROM a JOIN b ON a.col3 = b.col3`
- using IDSET will efficiently select out rowIDs from table A since there's no duplicates being process and roaring bitmap is super efficient in that 
  - this is super efficient if we don't have `b.col2` 
- however using IDSET also means at construction time of the bitmap, we lose the info on which row on b (specifically the column b.col2) should we tight back to the rows in table a after the filter
- weight pro-con: we should do a direct hash lookup if the cardinality of the selected b.col3 is high